core/trino-main/src/main/java/io/trino/metadata/DiscoveryNodeManager.java

```diff
@@ -208,81 +213,86 @@ public void refreshNodes()
         refreshNodesInternal();
     }
 
-    private synchronized void refreshNodesInternal()
+    private void refreshNodesInternal()
```

Comment:
I don't see any usage of InternalNodeManager in `io.trino.server.protocol.Query` and I'm not able to track the lock contention in JFR profiles back to this class.
So I'm not sure why any changes were needed here.
Please share rationale about this or consider dropping these changes if they don't have a sizeable impact.

Reply:
One execution of query tpcds/q64 is calling `TaskCountEstimator` more than 30k times which use `getAllNodes` which was synchronized, and it is blocked by `refreshNodesInternal`. It rather has minimal impact on blocking time during optimization, so I will drop this.

Reply:
We should rather look at the number of nodes in the cluster once at the start of planning for a query and then use that in task count calculations in all optimizer rules. It is probably not a good idea to let each optimizer rule use its own value of active worker nodes for task count estimation. This should be improved separately.
cc: @sopel39 @gaurav8297 @lukasz-stec